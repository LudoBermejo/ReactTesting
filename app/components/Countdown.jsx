import React from 'react';
import Clock from 'Clock';
import CountdownForm from 'CountdownForm';
import CountdownControls from 'CountdownControls';

export default class Countdown extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      totalSecondsToCount: 0,
      currentStatus: 'stopped'
    };
    this.startCountDown = this.startCountDown.bind(this);
    this.startTimer = this.startTimer.bind(this);
    this.renderFormOnlyIfStopped = this.renderFormOnlyIfStopped.bind(this);
    this.renderControlsIfNotStopped = this.renderControlsIfNotStopped.bind(this);
    this.timer = null;
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.state.currentStatus !== prevState.currentStatus) {
      switch (this.state.currentStatus) {
        case 'started':
          this.startTimer();
          break;
        case 'stopped':
          this.stopTimer();
          break;
        default: break;
      }
    }
  }

  stopTimer() {
    clearInterval(this.timer);
  }

  startTimer() {
    this.timer = setInterval(() => {
      let total = this.state.totalSecondsToCount;
      total = (total - 1 >= 0) ? total - 1 : 0;

      const state = {
        totalSecondsToCount: total
      };

      if (!total) {
        state.currentStatus = 'stopped';
      }

      this.setState(state);
    }, 1000);
  }

  startCountDown(seconds) {
    this.setState({
      totalSecondsToCount: seconds,
      currentStatus: 'started'
    });
  }

  renderFormOnlyIfStopped() {
    if (this.state.currentStatus === 'stopped') {
      return <CountdownForm onStartCountdown={this.startCountDown} />;
    }
    return '';
  }

  renderControlsIfNotStopped() {
    if (this.state.currentStatus !== 'stopped') {
      return (
        <CountdownControls
          countdownStatus={this.state.currentStatus}
          onClickPause={() => {}}
          onClickResume={() => {}}
          onClickStop={() => {}}
        />
      );
    }
    return '';
  }


  render() {
    return (
      <div>
        <Clock totalSeconds={this.state.totalSecondsToCount} />
        { this.renderFormOnlyIfStopped() }
        { this.renderControlsIfNotStopped(this.state.currentStatus) }
      </div>
    );
  }
}

