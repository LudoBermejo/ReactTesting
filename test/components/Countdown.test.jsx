import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import Countdown from 'Countdown';

describe('Countdown component', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Countdown onStartCountdown={() => {}} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('startTimer', () => {
    it('should set state to started', () => {
      const countdown = shallow(<Countdown onStartCountdown={() => {}} />);
      countdown.instance().startCountDown(5);
      expect(countdown.state().totalSecondsToCount).toBe(5);
      expect(countdown.state().currentStatus).toBe('started');
    });

    it('should set state to started and countdown', () => {
      jest.useFakeTimers();
      const countdown = shallow(<Countdown onStartCountdown={() => {}} />);
      countdown.instance().startCountDown(5);
      expect(countdown.state().totalSecondsToCount).toBe(5);
      expect(countdown.state().currentStatus).toBe('started');
      expect(setInterval.mock.calls.length).toBe(1);
      expect(setInterval.mock.calls[0][1]).toBe(1000);
      // Fast-forward until all timers have been executed
      jest.runOnlyPendingTimers();

      expect(countdown.state().totalSecondsToCount).toBe(4);
    });

    it('should stop when reach 0', () => {
      jest.useFakeTimers();
      const countdown = shallow(<Countdown onStartCountdown={() => {}} />);
      countdown.instance().startCountDown(1);
      expect(countdown.state().totalSecondsToCount).toBe(1);
      expect(countdown.state().currentStatus).toBe('started');
      expect(setInterval.mock.calls.length).toBe(1);
      expect(setInterval.mock.calls[0][1]).toBe(1000);
      // Fast-forward until all timers have been executed
      jest.runOnlyPendingTimers();

      expect(countdown.state().totalSecondsToCount).toBe(0);
      expect(countdown.state().currentStatus).toBe('stopped');
    });
  });
});
