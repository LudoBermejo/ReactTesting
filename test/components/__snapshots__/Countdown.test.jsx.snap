exports[`Countdown component renders correctly 1`] = `
<div>
  <div
    className="clock">
    <span
      className="clock-text">
      00:00
    </span>
  </div>
  <div>
    <form
      onSubmit={[Function]}>
      <input
        placeholder="Enter type in seconds"
        type="text" />
      <button
        className="button expanded"
        type="submit">
        Start
      </button>
    </form>
  </div>
  
</div>
`;
